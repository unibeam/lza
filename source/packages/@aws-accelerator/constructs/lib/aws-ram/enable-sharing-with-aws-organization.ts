/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

export interface EnableSharingWithAwsOrganizationProps {
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 * Class to initialize Policy
 */
export class EnableSharingWithAwsOrganization extends Construct {
  readonly id: string;

  constructor(scope: Construct, id: string, props: EnableSharingWithAwsOrganizationProps) {
    super(scope, id);

    const ENABLE_SHARING_WITH_AWS_ORGANIZATION_TYPE = 'Custom::EnableSharingWithAwsOrganization';

    //
    // Function definition for the custom resource
    //
    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, ENABLE_SHARING_WITH_AWS_ORGANIZATION_TYPE, {
      codeDirectory: path.join(__dirname, 'enable-sharing-with-aws-organization/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: [
            'ram:EnableSharingWithAwsOrganization',
            'iam:CreateServiceLinkedRole',
            'organizations:EnableAWSServiceAccess',
            'organizations:ListAWSServiceAccessForOrganization',
            'organizations:DescribeOrganization',
          ],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: ENABLE_SHARING_WITH_AWS_ORGANIZATION_TYPE,
      serviceToken: provider.serviceToken,
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
