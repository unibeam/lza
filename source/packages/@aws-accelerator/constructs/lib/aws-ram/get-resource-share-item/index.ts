/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
AWS.config.logger = console;

/**
 * get-resource-share-item - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Data: {
        arn: string;
      };
      Status: string;
    }
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const ramClient = new AWS.RAM({ customUserAgent: process.env['SOLUTION_ID'] });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const resourceOwner = event.ResourceProperties['resourceOwner'];
      const resourceShareArn = event.ResourceProperties['resourceShareArn'];
      const resourceType = event.ResourceProperties['resourceType'];

      let nextToken: string | undefined = undefined;
      do {
        const page = await throttlingBackOff(() =>
          ramClient
            .listResources({ resourceShareArns: [resourceShareArn], resourceType, resourceOwner, nextToken })
            .promise(),
        );
        // Return the first item found with the specified filters
        if (page.resources && page.resources.length > 0) {
          const item = page.resources[0];
          if (item.arn) {
            console.log(item.arn);
            return {
              PhysicalResourceId: item.arn.split('/')[1],
              Data: {
                arn: item.arn,
              },
              Status: 'SUCCESS',
            };
          }
        }
        nextToken = page.nextToken;
      } while (nextToken);

      throw new Error(`Resource share item not found`);

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
