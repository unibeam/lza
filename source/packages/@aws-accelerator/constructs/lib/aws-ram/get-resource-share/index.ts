/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * get-resource-share - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const ramClient = new AWS.RAM({ customUserAgent: process.env['SOLUTION_ID'] });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const resourceOwner = event.ResourceProperties['resourceOwner'];
      const owningAccountId = event.ResourceProperties['owningAccountId'];
      const name = event.ResourceProperties['name'];

      let nextToken: string | undefined = undefined;
      do {
        const page = await throttlingBackOff(() => ramClient.getResourceShareInvitations({ nextToken }).promise());
        for (const resourceShareInvitation of page.resourceShareInvitations ?? []) {
          if (
            resourceShareInvitation.status == 'PENDING' &&
            resourceShareInvitation.senderAccountId == owningAccountId &&
            resourceShareInvitation.resourceShareName === name
          ) {
            console.log(resourceShareInvitation);
            await throttlingBackOff(() =>
              ramClient
                .acceptResourceShareInvitation({
                  resourceShareInvitationArn: resourceShareInvitation.resourceShareInvitationArn!,
                })
                .promise(),
            );

            const found = await validateResourceShare(ramClient, resourceOwner, owningAccountId, name);

            if (found == false) {
              throw new Error(`Resource share ${name} not accepted successfully`); //share not found after multiple attempts
            }
          }
          nextToken = page.nextToken;
        }
      } while (nextToken);
      nextToken = undefined;
      do {
        const page = await throttlingBackOff(() =>
          ramClient.getResourceShares({ resourceOwner, nextToken: nextToken }).promise(),
        );
        for (const resourceShare of page.resourceShares ?? []) {
          if (resourceShare.owningAccountId == owningAccountId && resourceShare.name === name) {
            console.log(resourceShare);
            if (resourceShare.resourceShareArn) {
              return {
                PhysicalResourceId: resourceShare.resourceShareArn.split('/')[1],
                Status: 'SUCCESS',
              };
            }
          }
        }
        nextToken = page.nextToken;
      } while (nextToken);

      throw new Error(`Resource share ${name} not found`);

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}

async function validateResourceShare(
  ramClient: AWS.RAM,
  resourceOwner: string,
  owningAccountId: string,
  name: string,
): Promise<boolean | undefined> {
  let found = false;
  let counter = 5;
  do {
    const nextTokenShare: string | undefined = undefined;
    const pageResoureShares = await throttlingBackOff(() =>
      ramClient.getResourceShares({ resourceOwner, nextToken: nextTokenShare }).promise(),
    );
    if (pageResoureShares.resourceShares === undefined || pageResoureShares.resourceShares.length == 0) {
      await delay(5000); //delay 5 seconds and try again, no shares found
      console.log('resource share not found, waiting 5 seconds');
      counter = counter - 1;
      continue;
    } else {
      for (const resourceShare of pageResoureShares.resourceShares ?? []) {
        if (resourceShare.owningAccountId == owningAccountId && resourceShare.name === name) {
          found = true;
          break;
        }
      }
    }
    await delay(5000); //delay 5 seconds and try again, share not found
    console.log('resource share not found, waiting 5 seconds');
    counter = counter - 1;
  } while (found == false && counter >= 0);

  return found;
}

async function delay(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
