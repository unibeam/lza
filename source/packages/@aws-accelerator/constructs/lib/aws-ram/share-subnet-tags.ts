/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

export interface ShareSubnetTagsProps {
  readonly vpcTags?: cdk.CfnTag[];
  readonly subnetTags: cdk.CfnTag[];
  readonly sharedSubnetId: string;
  readonly owningAccountId: string;
  /**
   * Friendly name for the shared vpc
   */
  readonly vpcName: string;
  /**
   * Friendly name for the shared subnet
   */
  readonly subnetName: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly resourceLoggingKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
  /**
   * Accelerator SSM parameter Prefix
   */
  readonly acceleratorSsmParamPrefix: string;
}

/**
 * Class to initialize Policy
 */
export class ShareSubnetTags extends Construct {
  readonly id: string;

  constructor(scope: Construct, id: string, props: ShareSubnetTagsProps) {
    super(scope, id);

    const SHARE_SUBNET_TAGS_TYPE = 'Custom::ShareSubnetTags';

    //
    // Function definition for the custom resource
    //
    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, SHARE_SUBNET_TAGS_TYPE, {
      codeDirectory: path.join(__dirname, 'share-subnet-tags/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['ec2:DeleteTags', 'ec2:CreateTags'],
          Resource: [
            `arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:*:subnet/*`,
            `arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:*:vpc/*`,
          ],
        },
        {
          Effect: 'Allow',
          Action: ['ec2:DescribeTags', 'ec2:DescribeVpcs', 'ec2:DescribeSubnets'],
          Resource: '*',
        },
        {
          Effect: 'Allow',
          Action: ['ssm:GetParameter', 'ssm:PutParameter', 'ssm:DeleteParameter'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      properties: {
        vpcTags: props.vpcTags,
        subnetTags: props.subnetTags,
        sharedSubnetId: props.sharedSubnetId,
        sharedSubnetName: props.subnetName,
        vpcName: props.vpcName,
        acceleratorSsmParamPrefix: props.acceleratorSsmParamPrefix,
      },
      resourceType: SHARE_SUBNET_TAGS_TYPE,
      serviceToken: provider.serviceToken,
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.resourceLoggingKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
