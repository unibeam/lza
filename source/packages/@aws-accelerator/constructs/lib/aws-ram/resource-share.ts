/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import * as ram from 'aws-cdk-lib/aws-ram';
import { pascalCase } from 'change-case';
import { Construct } from 'constructs';
import { v4 as uuidv4 } from 'uuid';

const path = require('path');

export interface IResourceShare extends cdk.IResource {
  /**
   * The identifier of the resource share
   *
   * @attribute
   */
  readonly resourceShareId: string;

  /**
   * The name of the resource share
   *
   * @attribute
   */
  readonly resourceShareName: string;

  /**
   * The ARN of the resource share
   *
   * @attribute
   */
  readonly resourceShareArn: string;

  /**
   * The owner type of the resource share
   *
   * @attribute
   */
  readonly resourceShareOwner: ResourceShareOwner;
}

export enum ResourceShareOwner {
  SELF = 'SELF',
  OTHER_ACCOUNTS = 'OTHER-ACCOUNTS',
}

export interface ResourceShareLookupOptions {
  readonly resourceShareOwner: ResourceShareOwner;
  readonly resourceShareName: string;
  readonly owningAccountId?: string;
}

export interface ResourceShareProps {
  readonly name: string;
  readonly allowExternalPrincipals?: boolean;
  readonly permissionArns?: string[];
  readonly principals?: string[];
  readonly resourceArns?: string[];
}

export interface ResourceShareItemLookupOptions {
  readonly resourceShare: IResourceShare;
  readonly resourceShareItemType: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}
export interface IResourceShareItem extends cdk.IResource {
  readonly resourceShareItemArn: string;

  /**
   * The identifier of the shared resource item
   *
   * @attribute
   */
  readonly resourceShareItemId: string;
}

export abstract class ResourceShareItem extends cdk.Resource implements IResourceShareItem {
  readonly resourceShareItemArn: string = '';
  readonly resourceShareItemId: string = '';
  readonly resourceShareItemType: string = '';

  public static fromLookup(scope: Construct, id: string, options: ResourceShareItemLookupOptions): IResourceShareItem {
    class Import extends cdk.Resource implements IResourceShareItem {
      public readonly resourceShareItemArn: string;
      public readonly resourceShareItemId: string;

      constructor(scope: Construct, id: string) {
        super(scope, id);

        const GET_RESOURCE_SHARE_ITEM = 'Custom::GetResourceShareItem';

        const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, GET_RESOURCE_SHARE_ITEM, {
          codeDirectory: path.join(__dirname, 'get-resource-share-item/dist'),
          runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
          policyStatements: [
            {
              Effect: 'Allow',
              Action: ['ram:ListResources'],
              Resource: '*',
            },
          ],
        });

        const resource = new cdk.CustomResource(this, 'Resource', {
          resourceType: GET_RESOURCE_SHARE_ITEM,
          serviceToken: provider.serviceToken,
          properties: {
            resourceOwner: options.resourceShare.resourceShareOwner,
            resourceShareArn: options.resourceShare.resourceShareArn,
            resourceType: options.resourceShareItemType,
            uuid: uuidv4(), // Generates a new UUID to force the resource to update
          },
        });

        /**
         * Singleton pattern to define the log group for the singleton function
         * in the stack
         */
        const stack = cdk.Stack.of(scope);
        const logGroup =
          (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
          new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
            logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
            retention: options.logRetentionInDays,
            encryptionKey: options.kmsKey,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
          });
        resource.node.addDependency(logGroup);

        this.resourceShareItemId = resource.ref;
        this.resourceShareItemArn = resource.getAtt('arn').toString();
      }
    }
    return new Import(scope, id);
  }
}

/**
 * Creates a Resource Share
 */
export class ResourceShare extends cdk.Resource implements IResourceShare {
  public static fromLookup(scope: Construct, id: string, options: ResourceShareLookupOptions): IResourceShare {
    class Import extends cdk.Resource implements IResourceShare {
      public readonly resourceShareId: string;
      public readonly resourceShareName = options.resourceShareName;
      public readonly resourceShareOwner = options.resourceShareOwner;
      public readonly resourceShareArn;

      constructor(scope: Construct, id: string) {
        super(scope, id);
        const GET_RESOURCE_SHARE = 'Custom::GetResourceShare';

        //
        // Get the Resource Share Definition (by name)
        //
        const cr = cdk.CustomResourceProvider.getOrCreateProvider(this, GET_RESOURCE_SHARE, {
          codeDirectory: path.join(__dirname, 'get-resource-share/dist'),
          runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
          policyStatements: [
            {
              Effect: 'Allow',
              Action: ['ram:AcceptResourceShareInvitation', 'ram:GetResourceShareInvitations', 'ram:GetResourceShares'],
              Resource: '*',
            },
          ],
        });

        const resource = new cdk.CustomResource(this, 'Resource', {
          resourceType: GET_RESOURCE_SHARE,
          serviceToken: cr.serviceToken,
          properties: {
            name: options.resourceShareName,
            resourceOwner: options.resourceShareOwner,
            owningAccountId: options.owningAccountId,
            uuid: uuidv4(), // Generates a new UUID to force the resource to update
          },
        });

        this.resourceShareId = resource.ref;

        this.resourceShareArn = cdk.Stack.of(this).formatArn({
          service: 'ram',
          account: options.owningAccountId,
          resource: 'resource-share',
          arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
          resourceName: this.resourceShareId,
        });
      }
    }
    return new Import(scope, id);
  }

  public readonly resourceShareId: string;
  public readonly resourceShareName: string;
  public readonly resourceShareArn: string;
  public readonly resourceShareOwner: ResourceShareOwner;

  constructor(scope: Construct, id: string, props: ResourceShareProps) {
    super(scope, id);

    const resource = new ram.CfnResourceShare(this, pascalCase(`${props.name}ResourceShare`), {
      name: props.name,
      allowExternalPrincipals: props.allowExternalPrincipals,
      permissionArns: props.permissionArns,
      principals: props.principals,
      resourceArns: props.resourceArns,
    });

    this.resourceShareId = resource.ref;
    this.resourceShareName = props.name;
    this.resourceShareArn = resource.attrArn;
    this.resourceShareOwner = ResourceShareOwner.SELF;
  }
}
