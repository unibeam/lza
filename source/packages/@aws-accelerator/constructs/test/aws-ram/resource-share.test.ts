/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { ResourceShare, ResourceShareOwner, ResourceShareItem } from '../../lib/aws-ram/resource-share';
import { snapShotTest } from '../snapshot-test';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(ResourceShare): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new ResourceShare(stack, 'ResourceShare', {
  name: 'TestResourceShare',
  allowExternalPrincipals: true,
  permissionArns: [
    `arn:${stack.partition}:s3:::test-bucket-1-${stack.account}-${stack.region}`,
    `arn:${stack.partition}:s3:::test-bucket-2-${stack.account}-${stack.region}`,
  ],

  principals: ['accountID', 'organizationUnitId'],
  resourceArns: ['ec2:TransitGateway'],
});

const stackLookup = new cdk.Stack();

// Lookup resource share
ResourceShare.fromLookup(stackLookup, 'ResourceShareLookup', {
  resourceShareOwner: ResourceShareOwner.OTHER_ACCOUNTS,
  resourceShareName: 'ResourceShareName',
  owningAccountId: '111111111111',
});

/**
 * ResourceShare construct test
 */
describe('ResourceShare', () => {
  snapShotTest(testNamePrefix, stack);
});

//Lookup Resource share item
// const resourceShare: IResourceShare =
ResourceShareItem.fromLookup(stackLookup, 'ResourceShareItem', {
  logRetentionInDays: 7,
  kmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  resourceShareItemType: 'resourceShareItemType',
  resourceShare: ResourceShare.fromLookup(stackLookup, 'ResourceShareItemLookup', {
    resourceShareOwner: ResourceShareOwner.OTHER_ACCOUNTS,
    resourceShareName: 'ResourceShareName',
    owningAccountId: '111111111111',
  }),
});

/**
 * ResourceShare construct test
 */
describe('ResourceShareItem', () => {
  snapShotTest(testNamePrefix, stack);
});
