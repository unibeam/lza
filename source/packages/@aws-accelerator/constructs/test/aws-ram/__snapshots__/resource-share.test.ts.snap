// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceShare Construct(ResourceShare):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "ResourceShareTestResourceShareResourceShare8D7B67C7": {
      "Properties": {
        "AllowExternalPrincipals": true,
        "Name": "TestResourceShare",
        "PermissionArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":s3:::test-bucket-1-",
                {
                  "Ref": "AWS::AccountId",
                },
                "-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":s3:::test-bucket-2-",
                {
                  "Ref": "AWS::AccountId",
                },
                "-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
        ],
        "Principals": [
          "accountID",
          "organizationUnitId",
        ],
        "ResourceArns": [
          "ec2:TransitGateway",
        ],
      },
      "Type": "AWS::RAM::ResourceShare",
    },
  },
}
`;

exports[`ResourceShareItem Construct(ResourceShare):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "ResourceShareTestResourceShareResourceShare8D7B67C7": {
      "Properties": {
        "AllowExternalPrincipals": true,
        "Name": "TestResourceShare",
        "PermissionArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":s3:::test-bucket-1-",
                {
                  "Ref": "AWS::AccountId",
                },
                "-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":s3:::test-bucket-2-",
                {
                  "Ref": "AWS::AccountId",
                },
                "-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
        ],
        "Principals": [
          "accountID",
          "organizationUnitId",
        ],
        "ResourceArns": [
          "ec2:TransitGateway",
        ],
      },
      "Type": "AWS::RAM::ResourceShare",
    },
  },
}
`;
